File: ruby/lib/runapi/pixverse/resources/transition_video.rb
# frozen_string_literal: true

module RunApi
  module PixVerse
    module Resources
      # PixVerse transition-video resource.
      # Generate a transition between first and last frame images.
      class TransitionVideo
        include RunApi::Core::ResourceHelpers

        ENDPOINT = "/api/v1/pixverse/transition_video"

        RESPONSE_CLASS = Types::TransitionVideoResponse
        COMPLETED_RESPONSE_CLASS = Types::CompletedTransitionVideoResponse

        def initialize(http)
          @http = http
        end

        # Create a transition-video task and wait until complete.
        #
        # @param params [Hash] transition-video parameters
        # @return [RunApi::PixVerse::Types::CompletedTransitionVideoResponse] completed task with videos
        def run(options: nil, **params)
          task = create(options: options, **params)
          poll_until_complete { get(task.id, options: options) }
        end

        # Create a transition-video task.
        #
        # @param params [Hash] transition-video parameters
        # @return [RunApi::PixVerse::Types::TransitionVideoResponse] task creation result with id
        def create(options: nil, **params)
          params = compact_params(params)
          validate_params!(params)
          request(:post, ENDPOINT, body: params, options: options)
        end

        # Get transition-video task status by task ID.
        #
        # @param id [String] task ID
        # @return [RunApi::PixVerse::Types::TransitionVideoResponse] current task status
        def get(id, options: nil)
          request(:get, "#{ENDPOINT}/#{id}", options: options)
        end

        private

        def validate_params!(params)
          validate_contract!(CONTRACT["transition-video"], params)
        end
      end
    end
  end
end

File: python/src/runapi/pixverse/resources/transition_video.py
"""PixVerse transition-video resource."""

from typing import Any, Optional
from runapi.core import Resource, RequestOptions
from ..contract_gen import CONTRACT
from ..types import CompletedTransitionVideoResponse, TransitionVideoResponse

class TransitionVideo(Resource):
    ENDPOINT = "/api/v1/pixverse/transition_video"
    RESPONSE_CLASS = TransitionVideoResponse
    COMPLETED_RESPONSE_CLASS = CompletedTransitionVideoResponse
    def run(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        task = self.create(options=options, **params)
        return self._poll_until_complete(lambda: self.get(task.id, options=options))
    def create(self, options: Optional[RequestOptions] = None, **params: Any) -> Any:
        body = self._compact_params(params)
        self._validate_contract(CONTRACT["transition-video"], body)
        return self._request("post", self.ENDPOINT, body=body, options=options)
    def get(self, id: str, options: Optional[RequestOptions] = None) -> Any:
        return self._request("get", f"{self.ENDPOINT}/{id}", options=options)
